src/spinnaker.cpp

```diff
@@ -0,0 +1,1049 @@
+is_equal(const Trigger& lhs, const Trigger& rhs)
+{
+    return memcmp(&lhs, &rhs, sizeof(Trigger)) == 0;
+}
+
+// Returns true if a Spinnaker node is writable, false otherwise.
+bool
+check_node_writable(Spinnaker::GenApi::INode& node)
+{
+    if (!Spinnaker::GenApi::IsWritable(node)) {
+        LOGE("%s is not writable.", node.GetName().c_str());
+        return false;
+    }
+    return true;
+}
+
+// Various convenience Spinnaker node value setters.
+void
+set_enum_node(Spinnaker::GenApi::IEnumeration& node,
+              const Spinnaker::GenICam::gcstring& value)
+{
+    if (check_node_writable(node)) {
+        node = value;
+    }
+}
+
+void
+set_int_node(Spinnaker::GenApi::IInteger& node, int64_t value)
+{
+    if (check_node_writable(node)) {
+        const int64_t min = node.GetMin();
+        const int64_t max = node.GetMax();
+        value = std::clamp(value, min, max);
+        node = value;
+    }
+}
+
+void
+set_float_node(Spinnaker::GenApi::IFloat& node, double value)
+{
+    if (check_node_writable(node)) {
+        const double min = node.GetMin();
+        const double max = node.GetMax();
+        value = std::clamp(value, min, max);
+        node = value;
+    }
+}
+
+// Returns a representative Spinnaker binning node.
+// Spinnaker supports independent horizontal and vertical binning, but
+// Acquire only supports one binning value.
+Spinnaker::GenApi::IInteger&
+get_binning_node(const Spinnaker::CameraPtr& camera)
+{
+    // Favor a writable node if present.
+    Spinnaker::GenApi::IInteger& binning = camera->BinningVertical;
+    if (!IsWritable(camera->BinningVertical) &&
+        IsWritable(camera->BinningHorizontal)) {
+        binning = camera->BinningHorizontal;
+    }
+    return binning;
+}
+
+// Check that a cast from Acquire's camera ID is appropriate for Spinnaker.
+void
+check_spinnaker_camera_id(uint64_t id)
+{
+    constexpr uint64_t limit = std::numeric_limits<unsigned int>::max();
+    EXPECT(id < limit, "Expected an unsigned int device index. Got: %llu", id);
+}
+
+//
+// Camera declaration
+//
+
+struct SpinnakerCamera final : private Camera
+{
+    explicit SpinnakerCamera(Spinnaker::CameraPtr camera);
+    ~SpinnakerCamera();
+
+    void set(struct CameraProperties* properties);
+    void get(struct CameraProperties* properties);
+    void get_meta(struct CameraPropertyMetadata* meta) const;
+    void get_shape(struct ImageShape* shape) const;
+    void start();
+    void stop();
+    void execute_trigger() const;
+    void get_frame(void* im, size_t* nbytes, struct ImageInfo* info);
+
+  private:
+    // Guards access to all private state.
+    mutable std::mutex lock_;
+    // Used as a concise way to get and set a Spinnaker camera's state,
+    // rather than through its GenICam node map.
+    Spinnaker::CameraPtr camera_;
+    // Frame ID that starts at 0 each time the camera is started and
+    // increments for each frame retrieved.
+    // TODO: could use Spinnaker::Image::GetFrameID instead?
```

Comment:
I commented on this below but `GetFrameID` is preferred for the hardware frame id. We want whatever is closest to the camera hardware.

Reply:
Done.